### PR TITLE
Unregister legacy backends and disable tests

### LINKED ISSUE #622: Fully implement ignore_np_errstate for gtc:numpy
#620 Adds partial support for the option.

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Removes the legacy backends from the backend registery and thereby also disables running tests with these.

Resolves #622.

### CODE REVIEW

File: tests/test_unittest/test_call_interface.py [hunk 1]
```diff
@@ -91,50 +97,17 @@ def test_origin_selection():
     assert np.sum(np.asarray(C)) == 47
 
 
-@pytest.mark.parametrize("backend", INTERNAL_BACKENDS)
-def test_origin_k_fields(backend):
```
Comment (jdahm) [inline tests/test_unittest/test_call_interface.py:95]: I moved this to the integration tests.

File: src/gt4py/backend/gtc_backend/numpy/backend.py [hunk 2]
```diff
@@ -88,6 +88,8 @@ class GTCNumpyBackend(BaseBackend, CLIBackendMixin):
     name = "gtc:numpy"
     options: ClassVar[Dict[str, Any]] = {
         "oir_pipeline": {"versioning": True, "type": OirPipeline},
+        # TODO: Implement this option in source code
+        "ignore_np_errstate": {"versioning": True, "type": bool},
```
Comment (jdahm) [inline src/gt4py/backend/gtc_backend/numpy/backend.py:92]: I added the option `ignore_np_errstate` to the `gtc:numpy` backend (it didn't have it before), but I couldn't quickly figure out how to implement the feature cleanly.

File: tests/test_integration/test_code_generation.py [hunk 3]
```diff
@@ -361,7 +352,10 @@ def stencil(
             out_field = in_field * parameter
 
 
-@pytest.mark.parametrize("backend", OLD_BACKENDS)
+# TODO: Enable variable offsets on gtc:dace backend
+@pytest.mark.parametrize(
+    "backend", [backend for backend in ALL_BACKENDS if "dace" not in backend.values[0]]
+)
 def test_variable_offsets(backend):
```
Comment (jdahm) [inline tests/test_integration/test_code_generation.py:359]: Variable offsets work on all the new backends now (this test should have included them) with the exception of `gtc:dace`, so I made a TODO.

File: tests/test_integration/test_suites.py [hunk 4]
```diff
@@ -624,17 +623,7 @@ class TestNon3DFields(gt_testing.StencilTestSuite):
...
-        "numpy",
-        pytest.param("gtx86", marks=[pytest.mark.xfail]),
-        pytest.param("gtmc", marks=[pytest.mark.xfail]),
-        pytest.param("gtcuda", marks=[pytest.mark.xfail]),
-        "gtc:gt:cpu_ifirst",
-        "gtc:gt:cpu_kfirst",
-        "gtc:gt:gpu",
-        "gtc:dace",
-    ]
+    backends = ["gtc:gt:cpu_ifirst", "gtc:gt:cpu_kfirst", "gtc:gt:gpu", "gtc:dace"]
```
Comment (jdahm) [inline tests/test_integration/test_suites.py:626]: I tried setting this to simply `backends = INTERNAL_BACKENDS`, but it fails on `gtc:numpy`, so I made that a separate issue #621 and kept this.

File: src/gt4py/backend/gtc_backend/numpy/backend.py (hunk 2 above)
Comment (jdahm) [reply src/gt4py/backend/gtc_backend/numpy/backend.py:92]: It was required to add some sort of backend option to keep `test_backend_option_order` in test_cli.py.

**Comment by jdahm** [comment]
bors try

**Comment by jdahm** [comment]
bors try

**Comment by jdahm** [comment]
bors try

File: src/gt4py/backend/gtc_backend/numpy/backend.py (hunk 2 above)
Comment (egparedes) [reply src/gt4py/backend/gtc_backend/numpy/backend.py:92]: I might forget something, but I think to implement this feature only two changes are needed:

1. Pass the option to `NpirCodegen.apply` inside `GTCNumpyBackend.generate_computation()`:
```python
ignore_np_errstate = self.builder.options.backend_opts.get("ignore_np_errstate" , True)
source = NpirCodegen.apply(self.npir, ignore_np_errstate=ignore_np_errstate)
```
2. Modify the Jinja template in `NpirCodegen.Computation` to wrap all the code inside `run()` with a context handler to modify the numpy settings if ignore_np_errstate is `True` or an empty handler to keep the current settings if is `False`. Something like this (actual Jinja syntax _might_ be slightly different) :
```python
            def run({{ signature }}):
            {% if ignore_np_errstate %}
                 with np.errstate(divide='ignore', over='ignore', under='ignore', invalid='ignore'):
            {% else %}
                 with np.errstate():
            {% endif %}
                     ....
```

File: src/gtc/numpy/npir_codegen.py [hunk 5]
```diff
@@ -420,8 +420,14 @@ def run({{ signature }}):
...
 
+                {% if ignore_np_errstate %}
+                with np.errstate(divide='ignore', over='ignore', under='ignore', invalid='ignore'):
+                {% else %}
+                with np.errstate():
+                {% endif %}
+
                 {% for pass in vertical_passes %}
-                {{ pass | indent(4) }}
+                {{ pass | indent(8) }}
```
Comment (egparedes) [inline src/gtc/numpy/npir_codegen.py:430]:
```suggestion
                {{ pass | indent(8) }}
                {% else %}
                    pass
```

Comment (egparedes) [reply src/gtc/numpy/npir_codegen.py:430]: The other two failing tests errors are likely caused by the missing `ignore_err_state` kwarg in the `NpirCodegen.apply()` method. I guess the signature should be updated to reflect that this kwarg is compulsory (although this would likely break compatibility with the superclass for mypy)

File: src/gtc/cuir/kernel_fusion.py [hunk 6]
```diff
@@ -50,7 +50,8 @@ def is_parallel(kernel: cuir.Kernel) -> bool:
                 kernel.iter_tree()
                 .if_isinstance(cuir.FieldAccess)
                 .filter(
-                    lambda x: x.offset.i != 0 or x.offset.j != 0 or (x.offset.k != 0 and parallel)
+                    lambda x: any(off != 0 for off in x.offset.to_dict().values())
```
Comment (havogt) [inline src/gtc/cuir/kernel_fusion.py:53]: Why this change?

**Review by havogt — COMMENTED** [review]
Thanks @jdahm! Looks good, just one question.

File: src/gtc/cuir/kernel_fusion.py (hunk 6 above)
Comment (jdahm) [reply src/gtc/cuir/kernel_fusion.py:53]: It didn't work with VariableKOffset on a test that I enabled.

File: src/gt4py/backend/gt_backends.py [hunk 7]
```diff
@@ -717,11 +717,9 @@ def generate(self) -> Type["StencilObject"]:
             pyext_file_path=pyext_file_path,
         )
 
-    def generate_computation(self, *, ir: Any = None) -> Dict[str, Union[str, Dict]]:
-        if not ir:
-            ir = self.builder.implementation_ir
+    def generate_computation(self) -> Dict[str, Union[str, Dict]]:
         dir_name = f"{self.builder.options.name}_src"
-        src_files = self.make_extension_sources(ir=ir)
```
[the remark this answers is not in the document]
Comment (jdahm) [reply src/gt4py/backend/gt_backends.py:724]: Yeah that change in particular was interesting. There was a legacy backend that didn't actually conform with the abstract interface!

**Comment by jdahm** [comment]
bors try

**Comment by egparedes** [comment]
bors try

**Review by havogt — APPROVED** [review]
